pkg/processor/schemaregistry/client_test.go

```diff
@@ -0,0 +1,325 @@
+// Copyright © 2023 Meroxa, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package schemaregistry
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"sync"
+	"testing"
+
+	"github.com/conduitio/conduit/pkg/foundation/cerrors"
+	"github.com/conduitio/conduit/pkg/foundation/log"
+	"github.com/lovromazgon/franz-go/pkg/sr"
+	"github.com/matryer/is"
+)
+
+func TestClient_NotFound(t *testing.T) {
+	is := is.New(t)
+	ctx := context.Background()
+	logger := log.Nop()
+
+	rtr := newRoundTripRecorder(http.DefaultTransport)
+	c, err := NewClient(
+		logger,
+		sr.HTTPClient(&http.Client{Transport: rtr}),
+		sr.URLs(testSchemaRegistryURL(t)),
+	)
+	is.NoErr(err)
+
+	t.Run("SchemaByID", func(t *testing.T) {
+		is := is.New(t)
+
+		schema, err := c.SchemaByID(ctx, 12345)
+		is.True(err != nil)
+		is.Equal(sr.Schema{}, schema)
+
+		// check that error is expected
+		var respErr *sr.ResponseError
+		is.True(cerrors.As(err, &respErr))
+		is.Equal(40403, respErr.ErrorCode)
+
+		// check requests made by the client
+		is.Equal(len(rtr.Records()), 1)
+		rtr.AssertRecord(is, 0,
+			assertMethod("GET"),
+			assertRequestURI("/schemas/ids/12345"),
+			assertResponseStatus(404),
+			assertError(nil),
+		)
+	})
+
+	rtr.Clear() // clear requests between subtests
```

Comment:
I grouped the tests because they require the same setup. For these NotFound tests there is no setup (other than creating the client itself), so they could indeed be split, but I'd rather follow the same pattern as in the other tests to keep them consistent. I did however follow your advice on clearing the requests in a deferred statement in all subtests, this makes it a bit safer to add more subtests in the future.

Reply:
Yeah, most likely whoever is going to add any more tests is going to copy-paste or follow a similar patter at the very least.